cmd/rpc/sock.go

```diff
@@ -0,0 +1,455 @@
+package rpc
+
+import (
+	"fmt"
+	"github.com/canopy-network/canopy/controller"
+	"github.com/canopy-network/canopy/lib"
+	"github.com/gorilla/websocket"
+	"github.com/julienschmidt/httprouter"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+)
+
+/* This file implements the client & server logic for the 'root-chain info' and corresponding 'on-demand' calls to the rpc */
+
+var _ lib.RCManagerI = new(RCManager)
+
+// RCManager handles a group of root-chain sock clients
+type RCManager struct {
+	c             lib.Config                   // the global node config
+	subscriptions map[uint64]*RCSubscription   // chainId -> subscription
+	subscribers   map[uint64]*RCSubscriber     // chainId -> subscriber
+	l             sync.Mutex                   // thread safety
+	afterRCUpdate func(info lib.RootChainInfo) // callback after the root chain info update
+	upgrader      websocket.Upgrader           // upgrade http connection to ws
+	log           lib.LoggerI                  // stout log
+}
+
+// NewRCManager() constructs a new instance of a RCManager
+func NewRCManager(controller *controller.Controller, c lib.Config, l lib.LoggerI) (man *RCManager) {
+	// create the manager
+	man = &RCManager{
+		c:             c,
+		subscriptions: make(map[uint64]*RCSubscription),
+		subscribers:   make(map[uint64]*RCSubscriber),
+		l:             controller.Mutex,
```

Comment:
why a mutex from the controller? from the commit I see it fixed the deadlock issue but is odd from a new user why would you use an external mutex, could you add a simple comment explaining why?

Reply:
It's related to https://github.com/canopy-network/canopy/issues/157

But ultimately, the pattern locking the `controller.Mutex` is the safest way to avoid multiple locks causing deadlocks.